src/main/java/org/folio/rest/transaction/PaymentCreditHandler.java

```diff
@@ -0,0 +1,373 @@
+      //update all the re-calculated encumbrances into the Transaction table
+      .map(map -> new Tx<>(map.values()
+        .stream()
+        .collect(Collectors.toList()), tx.getPgClient()).withConnection(tx.getConnection()))
+      .compose(this::updatePermanentTransactions)
+      .compose(ok -> Future.succeededFuture(tx));
+
+  }
+
+  /**
+   * <pre>
+   * Encumbrances are recalculated with the credit transaction as below:
+   * - expended decreases by credit transaction amount (min 0)
+   * - amount (effectiveEncumbrance) is recalculated= (initialAmountEncumbered - (amountAwaitingPayment + amountExpended))
+   * </pre>
+   */
+  private Map<String, Transaction> applyCredits(List<Transaction> tempTxns, Map<String, Transaction> encumbrancesMap) {
+
+    List<Transaction> tempCredits = tempTxns.stream()
+      .filter(isCreditTransaction.and(hasEncumbrance))
+      .collect(Collectors.toList());
+
+    if (tempCredits.isEmpty()) {
+      return encumbrancesMap;
+    }
+    CurrencyUnit currency = Monetary.getCurrency(tempCredits.get(0)
+      .getCurrency());
+    tempCredits.forEach(creditTxn -> {
+      Transaction encumbranceTxn = encumbrancesMap.get(creditTxn.getPaymentEncumbranceId());
+
+      double newExpended = MoneyUtils.subtractMoneyNonNegative(encumbranceTxn.getEncumbrance()
+        .getAmountExpended(), creditTxn.getAmount(), currency);
+      encumbranceTxn.getEncumbrance()
+        .setAmountExpended(newExpended);
+
+      // recalculate effectiveEncumbrance ( initialAmountEncumbered - (amountAwaitingPayment + amountExpended))
+      double newAmount = MoneyUtils.subtractMoney(encumbranceTxn.getEncumbrance()
+        .getInitialAmountEncumbered(),
+          MoneyUtils.sumMoney(encumbranceTxn.getEncumbrance()
+            .getAmountAwaitingPayment(),
+              encumbranceTxn.getEncumbrance()
+                .getAmountExpended(),
+              currency),
+          currency);
+      encumbranceTxn.setAmount(newAmount);
+    });
+
+    return encumbrancesMap;
+  }
+
+  /**
+   * <pre>
+   * Encumbrances are recalculated with the payment transaction as below:
+   * - awaitingPayment decreases by payment transaction amount
+   * - expended increases by payment transaction amount
+   * - amount decreases by the payment transaction amount
+   * </pre>
+   */
+  private Map<String, Transaction> applyPayments(List<Transaction> tempTxns, Map<String, Transaction> encumbrancesMap) {
+    List<Transaction> tempPayments = tempTxns.stream()
+      .filter(isPaymentTransaction.and(hasEncumbrance))
+      .collect(Collectors.toList());
+    if (tempPayments.isEmpty()) {
+      return encumbrancesMap;
+    }
+
+    CurrencyUnit currency = Monetary.getCurrency(tempPayments.get(0)
+      .getCurrency());
+    tempPayments.forEach(pymtTxn -> {
+      Transaction encumbranceTxn = encumbrancesMap.get(pymtTxn.getPaymentEncumbranceId());
+      double newAwaitingPayment = MoneyUtils.subtractMoney(encumbranceTxn.getEncumbrance()
```

Comment:
NPE may arise here if the encumbrance with `id=pymtTxn.getPaymentEncumbranceId()` does not exist.

Reply:
same for applyCredits

Reply:
That condition is handled in https://github.com/folio-org/mod-finance-storage/blob/81326caa16909722024f8c5089a8a86d1a775dc8/src/main/java/org/folio/rest/transaction/PaymentCreditHandler.java#L200, it is considered a payment only if paymentEncumbranceId is not null

Reply:
I mean the situation when the `paymentEncumbranceId` is specified, but there is no such encumbrance in the database.

Reply:
It's a valid use case for a payment/credit to NOT have an encumbrance associated with it... These still need to be applied.

Reply:
The payments/credits that do not have an encumbrance are handled separately here https://github.com/folio-org/mod-finance-storage/blob/81326caa16909722024f8c5089a8a86d1a775dc8/src/main/java/org/folio/rest/transaction/PaymentCreditHandler.java#L143

Reply:
> I mean the situation when the `paymentEncumbranceId` is specified, but there is no such encumbrance in the database.

There is no possibility for that , as there is a foreign key relationship on paymentEncumbranceId on transaction table. A 400 error would be returned if we try to insert such a record
https://github.com/folio-org/mod-finance-storage/blob/05aac96ca596da1c1a21f735601dbe4c8976c4bb/src/main/resources/templates/db_scripts/schema.json#L480

Reply:
Also added a test case to cover this scenario